olive/evaluator/olive_evaluator.py

```diff
@@ -312,6 +314,49 @@ def format_input(input_data, io_config):
             if k in input_names
         }
 
+    @staticmethod
+    def prepare_io_bindings(session, input_data, device, device_id: int = 0, shared_kv_buffer: bool = False):
+        """Convert input from numpy array to OrtValue.
+
+        session: ONNXRuntime session
+        input_data: dict of input data, value is numpy array
+        device: olive device
+        device_id: 0 by default. TODO(trajep): support user to specified device id
+        shared_kv_buffer: whether to share the key/value buffer across multiple runs, it is False by default,
+            and only used when we observe kv cache and fp16 is used.
+            TODO(trajep): how shared_kv_buffer works with generation task
+        """
+        from onnxruntime import OrtValue
+
+        use_fp16 = any(v.dtype == np.float16 for v in input_data.values())
+        io_bind_op = session.io_binding()
+        io_bind_device = "cuda" if device == "gpu" else "cpu"
+
+        if shared_kv_buffer:
+            OnnxEvaluator.KV_CACHE_ORTVALUES = OnnxEvaluator.KV_CACHE_ORTVALUES or {}
+
+        for k, v in input_data.items():
+            if shared_kv_buffer and use_fp16 and ("cache" in k or "past_key_values" in k):
+                if k not in OnnxEvaluator.KV_CACHE_ORTVALUES:
```

Comment:
good catch for memory sake

Reply:
Emm, I think the shared buffer is something like to pin a piece of GPU memory to accelerate generation(token) task.  That is why I added: `TODO(trajep): how shared_kv_buffer works with generation task in olive`

Note that, there is the operation to bind ortvalue output into shared buffer. Then actually I think in our latency measurement, the input should be chagned.
`io_bind_op.bind_ortvalue_output(name, OnnxEvaluator.KV_CACHE_ORTVALUES[input_name])`


We need a central attribute to ensure it is shared among multi-running. Maybe class attributes is too broader. Then how about instance attribute?

Reply:
For binding the output to the input, we don't need a central place to store the cache. The `prepare_io_binding` method can have a local `cache` dict for it. 
The cache only comes into play if we want to run the same model with different inputs. Even then, it depends on whether the gpu memory gets freed when the `io_bind_op` object goes out of scope. If it does, there is no need for a central cache since it gets refreshed at each iteration of the loop (over dataloader). 
The only advantage would could be speed up over the io_binding process itself since we won't need to allocate memory every time. If we do this, we could keep a cache dict inside `_evaluate_onnx_latency`.

Reply:
Let us not limit in Olive latency measurement firstly, the key advantage to enable shared buffer is to avoid memory allocation/release in every forward call. The kv cache should be updated_inplace since 2rd forward call. (actually shared buffer should be able cover all the transformers kv cache case not only GQA fp16.) 
For this goal, I think we are on the same line.

Actually from my side, I do not think shared buffer is required in our latency measurement.  It is more like the part of generation predication.

Then it comes to latency measurement, what if we enable shared buffer in **Olive**:
1. w/o shared buffer: the input are always be same including the kv cache.
2. with shared buffer: the input_ids, attention masks, position id are same, but kv cache got refreshed as it bind ortvalue output to shared buffer. 
Right?


But anyway in order to support the shared buffer in Olive(at least we have the capacity in Olive), we can leverage latency as a test. 
I would be agree the idea of `we could keep a cache dict inside _evaluate_onnx_latency.`
Like:

```python
def _evaluate_onnx_latency(...):
    ...
    kv_cache_buffer = {}
    prepare_io_bindings(..., kv_cache_buffer)
    ....
    del kv_cache
```